algebra/euclidean_domain.lean

```diff
@@ -0,0 +1,316 @@
+/- 
+Copyright (c) 2018 Louis Carlin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Louis Carlin
+
+Euclidean domains and Euclidean algorithm (extended to come)
+A lot is based on pre-existing code in mathlib for natural number gcds
+-/
+import tactic.ring
+
+universe u
+
+class euclidean_domain (α : Type u) extends integral_domain α :=
+(quotient : α → α → α)
+(remainder : α → α → α)
+(witness : ∀ a b, (quotient a b) * b + (remainder a b) = a) -- This could be changed to the same order as int.mod_add_div. We normally write qb+r rather than r + qb though.
+(valuation : α → ℕ)
+(val_mod : ∀ a b, b = 0 ∨  valuation (remainder a b) <  valuation b)
+(val_mul : ∀ a b, b = 0 ∨ valuation a ≤ valuation (a*b))
+(decidable_eq : ∀ a b : α, decidable (a=b))
```

Comment:
I was thinking more about this, and I don't think it is a good idea to add `decidable_eq` as a structure field. The problem comes when you actually want to apply it to a (semi) concrete decidable type.

For example for `A * B` we would have suddenly two ways to show decidability:
  `euclidean_domain.decidable_eq`
and 
  `prod.decidable_eq`.

`decidable` is special, as it is a subsingleton. But this is not known to the type class mechanism.

But even worse, I think we might get into problems with classes in the type class hierarchy after `integral_domain`, i.e. `discrete_field` and `field`.

I think we should write it as:
```lean
class euclidean_domain (α : Type u) [integral_domain α] [decidable_eq α]:=
...
```

Reply:
I've changed to ```class euclidean_domain (α : Type u) [decidable_eq α] extends integral_domain α```.

Could you elaborate on the problems which occur with `discrete_field` and `field` from extending `integral_domain`?
Does the `(has_decidable_eq : decidable_eq α)` field of `discrete_field` cause the same problems with products?